android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+
+                // Apply message style only for report comments
+                if (REPORT_COMMENT_TYPE.equals(payload.get(TYPE_KEY).getString())) {
+                        applyMessageStyle(builder, payload, arguments.getNotificationId());
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Failed to parse conversation", e);
+            }
+        }
+
+        return builder;
+    }
+
+    /**
+     * Applies the message style to the notification builder. It also takes advantage of the
+     * notification cache to build conversations.
+     *
+     * @param builder Notification builder that will receive the message style
+     * @param payload Notification payload, which contains all the data we need to build the notifications.
+     * @param notificationID Current notification ID
+     */
+    private void applyMessageStyle(NotificationCompat.Builder builder, JsonMap payload, int notificationID) {
+        int reportID = payload.get("reportID").getInt(-1);
+        NotificationCache notificationCache = findOrCreateNotificationCache(reportID);
+        JsonMap reportAction = payload.get("reportAction").getMap();
+        String name = reportAction.get("person").getList().get(0).getMap().get("text").getString();
+        String avatar = reportAction.get("avatar").getString();
+        String accountID = Integer.toString(reportAction.get("actorAccountID").getInt(-1));
+        String message = reportAction.get("message").getList().get(0).getMap().get("text").getString();
+        long time = reportAction.get("timestamp").getLong(0);
+        String roomName = payload.get("roomName") == null ? "" : payload.get("roomName").getString("");
+        String conversationTitle = "Chat with " + name;
+        if (!roomName.isEmpty()) {
+            conversationTitle = "#" + roomName;
+        }
+
+        IconCompat iconCompat = fetchIcon(avatar, FALLBACK_ICON_ID);
+        Person person = new Person.Builder()
+                .setIcon(iconCompat)
+                .setKey(accountID)
+                .setName(name)
+                .build();
+
+        if (!notificationCache.people.containsKey(accountID)) {
+            notificationCache.people.put(accountID, person);
+        }
+
+        notificationCache.messages.add(new NotificationCache.Message(person, message, time));
+
+        NotificationCompat.MessagingStyle messagingStyle = new NotificationCompat.MessagingStyle(person)
+                .setGroupConversation(notificationCache.people.size() > 2 || !roomName.isEmpty())
+                .setConversationTitle(conversationTitle);
+
+        for (NotificationCache.Message cachedMessage : notificationCache.messages) {
+            messagingStyle.addMessage(cachedMessage.text, cachedMessage.time, cachedMessage.person);
+        }
+
+        if (notificationCache.prevNotificationID != -1) {
+            NotificationManagerCompat.from(context).cancel(notificationCache.prevNotificationID);
+        }
```

Comment:
NAB: There's a lot going on in this function that isn't obvious without reading all the code. A few comments would help

Reply:
Added!